src/Conversion/ONNXToKrnl/Math/Elementwise.cpp

```diff
@@ -331,6 +334,74 @@ struct ScalarOp<ONNXTanOp> {
   using IOp = NotSuportedScalarOp;
 };
 
+//===----------------------------------------------------------------------===//
+// Scalar unary ops for lowering ONNXGeluOp
+//===----------------------------------------------------------------------===//
+template <>
+struct ScalarOp<ONNXGeluOp> {
+  using FOp = CustomScalarOp;
+  using IOp = CustomScalarOp;
+};
+
+template <>
+double analyzeSimdFor<ONNXGeluOp>(
+    Type t, Operation *op, int64_t &von, int64_t &son) {
+  double results;
+  StringRef approximate = dyn_cast<ONNXGeluOp>(op).getApproximate();
+  if (approximate == "none")
+    results = simdAnalysis({GenericOps::ArithmeticGop, GenericOps::DivGop,
+                               GenericOps::MulGop, GenericOps::SqrtGop},
+        {1, 1, 2, 1}, t, von, son);
+  if (approximate == "tanh")
+    results = simdAnalysis(
+        {GenericOps::ArithmeticGop, GenericOps::DivGop, GenericOps::MulGop,
+            GenericOps::PowGop, GenericOps::SqrtGop},
+        {2, 1, 4, 1, 1}, t, von, son);
+  return results;
+}
+
+template <>
+Value emitScalarOpFor<ONNXGeluOp>(ConversionPatternRewriter &rewriter,
+    Location loc, Operation *op, Type elementType,
+    ArrayRef<Value> scalarOperands) {
+  Value operand = scalarOperands[0];
+
+  CheckIfCustomScalarOpIsSupported<ONNXGeluOp>(elementType);
+  MultiDialectBuilder<KrnlBuilder, MathBuilder> create(rewriter, loc);
+
+  // Approximate is a required attribute and should have a default value of
+  // "none". "approximate = none" simply implies no approximation will take
+  // place. However, "approximate" can also have a string value of "tanh" which
+  // indicates the use of tanh approximation.
+  StringRef approximate = dyn_cast<ONNXGeluOp>(op).getApproximate();
+
+  // Create constants
+  Value half = create.math.constant(elementType, 0.5);
+  Value one = create.math.constant(elementType, 1);
+  Value three = create.math.constant(elementType, 3);
+  Value decimal = create.math.constant(elementType, 0.044715);
+  Value sqrt_two_over_pi = create.math.constant(elementType, sqrt(2 / M_PI));
+  Value sqrt_two = create.math.constant(elementType, sqrt(2));
+
+  // Calculations
+  Value div = create.math.div(operand, sqrt_two);
+  Value erfApprox = create.math.erf(div);
+  Value add = create.math.add(one, erfApprox);
+  Value dec = create.math.add(
+      operand, create.math.mul(decimal, create.math.pow(operand, three)));
+  Value tanhApprox = create.math.tanh(create.math.mul(sqrt_two_over_pi, dec));
+  // Approximate = none returns an output of y = 0.5 * x * (1 +
+  // erf(x/sqrt(2)))
+  if (approximate == "none")
+    return create.math.mul(create.math.mul(half, operand), add);
+  // Approximate = tanh returns an output of y = 0.5 * x * (1 + Tanh(sqrt(2/pi)
+  // * (x + 0.044715 * x^3)))
+  if (approximate == "tanh")
+    return create.math.mul(
```

Comment:
Same here, please move local variables `tanhApprox`, `dec`,  ... into the body of this `if`:
```
  Value three = create.math.constant(elementType, 3);
  Value decimal = create.math.constant(elementType, 0.044715);
  Value sqrt_two_over_pi = create.math.constant(elementType, sqrt(2 / M_PI));
  Value dec = create.math.add(operand, create.math.mul(decimal, create.math.pow(operand, three)));
  Value tanhApprox = create.math.tanh(create.math.mul(sqrt_two_over_pi, dec));
```

Reply:
Okay when I move the Value's into the `if` I actually get the following errors:

```
[10/14] Building CXX object src/Conversion/ONNXToKrnl/CMakeFiles/OMONNXToKrnl.dir/Math/Elementwise.cpp.o
FAILED: src/Conversion/ONNXToKrnl/CMakeFiles/OMONNXToKrnl.dir/Math/Elementwise.cpp.o 
/usr/bin/c++ -DONNX_ML=1 -DONNX_MLIR_DECOMP_ONNX_CONVTRANSPOSE -DONNX_MLIR_ENABLE_STABLEHLO -DONNX_NAMESPACE=onnx -DONNX_USE_LITE_PROTO=1 -D_DEBUG -D_GLIBCXX_ASSERTIONS -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/Users/meganhampton/zDLC/llvm-project/llvm/include -I/Users/meganhampton/zDLC/llvm-project/build/include -I/Users/meganhampton/zDLC/llvm-project/mlir/include -I/Users/meganhampton/zDLC/llvm-project/build/tools/mlir/include -I/Users/meganhampton/zDLC/onnx-mlir -I/Users/meganhampton/zDLC/onnx-mlir/build -I/Users/meganhampton/zDLC/onnx-mlir/include -I/Users/meganhampton/zDLC/onnx-mlir/third_party/onnx -I/Users/meganhampton/zDLC/onnx-mlir/build/third_party/onnx -isystem /opt/homebrew/include -isystem /opt/homebrew/Cellar/protobuf@21/21.12/include -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -DSUPPRESS_THIRD_PARTY_WARNINGS -g -std=gnu++17 -arch arm64 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX14.2.sdk   -D_DEBUG -D_GLIBCXX_ASSERTIONS -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -MD -MT src/Conversion/ONNXToKrnl/CMakeFiles/OMONNXToKrnl.dir/Math/Elementwise.cpp.o -MF src/Conversion/ONNXToKrnl/CMakeFiles/OMONNXToKrnl.dir/Math/Elementwise.cpp.o.d -o src/Conversion/ONNXToKrnl/CMakeFiles/OMONNXToKrnl.dir/Math/Elementwise.cpp.o -c /Users/meganhampton/zDLC/onnx-mlir/src/Conversion/ONNXToKrnl/Math/Elementwise.cpp
/Users/meganhampton/zDLC/onnx-mlir/src/Conversion/ONNXToKrnl/Math/Elementwise.cpp:390:5: warning: misleading indentation; statement is not part of the previous 'if' [-Wmisleading-indentation]
    Value erfApprox = create.math.erf(div);
    ^
/Users/meganhampton/zDLC/onnx-mlir/src/Conversion/ONNXToKrnl/Math/Elementwise.cpp:388:3: note: previous statement is here
  if (approximate == "none")
  ^
/Users/meganhampton/zDLC/onnx-mlir/src/Conversion/ONNXToKrnl/Math/Elementwise.cpp:389:11: warning: unused variable 'div' [-Wunused-variable]
    Value div = create.math.div(operand, sqrtTwo);
          ^
/Users/meganhampton/zDLC/onnx-mlir/src/Conversion/ONNXToKrnl/Math/Elementwise.cpp:390:39: error: no viable conversion from '<overloaded function type>' to 'mlir::Value'
    Value erfApprox = create.math.erf(div);
                                      ^~~
/Users/meganhampton/zDLC/llvm-project/mlir/include/mlir/IR/Value.h:96:7: note: candidate constructor (the implicit copy constructor) not viable: no overload of 'div' matching 'const mlir::Value &' for 1st argument
class Value {
      ^
/Users/meganhampton/zDLC/llvm-project/mlir/include/mlir/IR/Value.h:96:7: note: candidate constructor (the implicit move constructor) not viable: no overload of 'div' matching 'mlir::Value &&' for 1st argument
/Users/meganhampton/zDLC/llvm-project/mlir/include/mlir/IR/Value.h:98:13: note: candidate constructor not viable: no overload of 'div' matching 'detail::ValueImpl *' for 1st argument
  constexpr Value(detail::ValueImpl *impl = nullptr) : impl(impl) {}
            ^
/Users/meganhampton/zDLC/onnx-mlir/src/Dialect/Mlir/DialectBuilder.hpp:116:31: note: passing argument to parameter 'val' here
  mlir::Value erf(mlir::Value val) const;
                              ^
/Users/meganhampton/zDLC/onnx-mlir/src/Conversion/ONNXToKrnl/Math/Elementwise.cpp:397:5: warning: misleading indentation; statement is not part of the previous 'if' [-Wmisleading-indentation]
    Value tanhApprox = create.math.tanh(create.math.mul(sqrtTwoOverPi, dec));
    ^
/Users/meganhampton/zDLC/onnx-mlir/src/Conversion/ONNXToKrnl/Math/Elementwise.cpp:395:3: note: previous statement is here
  if (approximate == "tanh")
  ^
/Users/meganhampton/zDLC/onnx-mlir/src/Conversion/ONNXToKrnl/Math/Elementwise.cpp:396:11: warning: unused variable 'dec' [-Wunused-variable]
    Value dec = create.math.add(operand, create.math.mul(decimal, create.math.pow(operand, three)));
          ^
/Users/meganhampton/zDLC/onnx-mlir/src/Conversion/ONNXToKrnl/Math/Elementwise.cpp:397:72: error: use of undeclared identifier 'dec'
    Value tanhApprox = create.math.tanh(create.math.mul(sqrtTwoOverPi, dec));
                                                                       ^
4 warnings and 2 errors generated.
```

Reply:
Fixed